lib/shoulda/matchers/active_record/have_encrypt_attribute_matcher.rb

```diff
@@ -0,0 +1,160 @@
+        #     end
+        #
+        #     # Minitest (Shoulda)
+        #     class SurveyTest < ActiveSupport::TestCase
+        #       should have_encrypt_attribute(:access_code).downcase(true)
+        #     end
+        #
+        # ##### ignore_case
+        #
+        #     class Survey < ActiveRecord::Base
+        #       encrypts :access_code, deterministic: true, ignore_case: true
+        #     end
+        #
+        #     # RSpec
+        #     RSpec.describe Survey, type: :model do
+        #       it { should have_encrypt_attribute(:access_code).ignore_case(true) }
+        #     end
+        #
+        #     # Minitest (Shoulda)
+        #     class SurveyTest < ActiveSupport::TestCase
+        #       should have_encrypt_attribute(:access_code).ignore_case(true)
+        #     end
+        #
+        # @return [HaveEncryptAttributeMatcher]
+        #
+        def have_encrypt_attribute(value)
+            HaveEncryptAttributeMatcher.new(value)
+        end
+  
+        # @private
+        class HaveEncryptAttributeMatcher
+          def initialize(attribute)
+            @attribute = attribute.to_sym
+            @options = {}
+          end
+  
+          attr_reader :failure_message, :failure_message_when_negated
+          
+          def deterministic(deterministic)
+            @options[:deterministic] = deterministic
+            self
+          end
+
+          def downcase(downcase)
+            @options[:downcase] = downcase
+            self
+          end
+
+          def ignore_case(ignore_case)
+            @options[:ignore_case] = ignore_case
+            self
+          end
+
+          def matches?(subject)
+            @subject = subject
+            encrypted_attributes_included? &&
+              options_correct?(
+                :deterministic, 
+                :downcase, 
+                :ignore_case
+              )
+          end
+  
+          def description
+            "encrypt #{@attribute}"
+          end
+  
+          private
+
+          def encrypted_attributes_included?
+            if encrypted_attributes.include?(@attribute)
+              @failure_message_when_negated = "Did not expect #{@attribute} to be encrypted"
+              true
+            else
+              @failure_message = "Expect #{@attribute} should be encrypted, but not"
+              false
+            end
+          end
+
+          def options_correct?(*opts)
+            opts.each do |opt|
+              if @options.has_key?(opt)
+                expected = @options[opt]
+                actual = encrypted_attribute_scheme.instance_variable_get("@#{opt}")
+                if expected == actual
+                  true
+                else
+                  @failure_message = "#{opt} option should be #{expected}, got #{actual}"
+                  return false
+                end
+              else
+                true
+              end
+            end
+            true
+          end

```

Comment:
I think we could simplify that method using guard clauses it feels a little confusing with all the `true`, and `false` values being returned. Another small improvement would be getting the value from the `encrypted_attribute_scheme` using the predicate public methods of that class. https://github.com/rails/rails/blob/900dc3280932df4a614cfa06b59959f249e65ce1/activerecord/lib/active_record/encryption/scheme.rb#L31-L40

On the other hand, I think we could also adjust the `failure_message` to be more explicit.

```suggestion
          def options_correct?(*opts)
            opts.all? do |opt|
              next true unless @options.key?(opt)

              expected = @options[opt]
              actual = encrypted_attribute_scheme.public_send("#{opt}?")

              next true if expected == actual

              @failure_message = "Expected to #{description} of #{class_name} using :#{opt} option as ‹#{expected}›, but got ‹#{actual}›"
              false
            end
          end
```

This change would require some changes in the specs, but I think it's a nice improvement, what do you think?

Reply:
it's better!